drivers/ata/ahciem.c

```diff
@@ -0,0 +1,566 @@
+			enc->status[i][0] = 0;
+			enc->status[i][1] = ads[1] & 0x02;		/* rebuild/remap */
+			enc->status[i][2] = ads[2] & (0x80 | 0x02);	/* rqst active + rqst ident */
+			enc->status[i][3] = ads[3] & 0x20;		/* rqst fault */
+			ahciem_setleds(enc, i);
+		} else if (ads0[0] & 0x80) {
+			enc->status[i][0] = 0;
+			enc->status[i][1] = ads0[1] & 0x02;		/* rebuild/remap */
+			enc->status[i][2] = ads0[2] & (0x80 | 0x02);	/* rqst active + rqst ident */
+			enc->status[i][3] = ads0[3] & 0x20;		/* rqst fault */
+			ahciem_setleds(enc, i);
+		}
+	}
+
+	kfree(page);
+}
+
+static int ahciem_queuecommand(struct Scsi_Host *shost, struct scsi_cmnd *cmd)
+{
+	struct ahciem_enclosure *enc = (struct ahciem_enclosure *)&shost->hostdata[0];
+	struct ahciem_args args = { .cmd = cmd, .enc = enc };
+	const u8 *cdb = cmd->cmnd;
+
+	if (unlikely(!cmd->cmd_len)) {
+		printk("ahciem: bad CDB len=0 in %s\n", DRV_NAME);
+		cmd->result = DID_ERROR << 16;
+		scsi_done(cmd);
+		return 0;
+	}
+
+	/*
+	 * Commands required for SES devices by SPC:
+	 *  - INQUIRY
+	 *  - REPORT LUNS
+	 *  - REQUEST SENSE
+	 *  - TEST UNIT READY
+	 *  - SEND DIAGNOSTIC
+	 *  - RECEIVE DIAGNOSTIC RESULTS
+	 */
+	switch (cdb[0]) {
+	case INQUIRY:
+		if (cdb[1] & 2)			/* obsolete CMDDT set? */
+			ahciem_scsi_set_invalid_field(cmd, 1, 1, true);
+		else if ((cdb[1] & 1) == 0) {	/* standard INQUIRY */
+			if (cdb[2] != 0)
+				ahciem_scsi_set_invalid_field(cmd, 2, 0xff, true);
+			else
+				ahciem_rbuf_fill(&args, ahciem_scsiop_inq_std);
+		} else switch (cdb[2]) {	/* VPD INQUIRY */
+		case 0x00:	/* Supported VPD Pages */
+			ahciem_rbuf_fill(&args, ahciem_scsiop_inq_00);
+			break;
+		case 0x83:	/* Device Identification */
+			ahciem_rbuf_fill(&args, ahciem_scsiop_inq_83);
+			break;
+		default:
+			ahciem_scsi_set_invalid_field(cmd, 2, 0xff, true);
+			break;
+		}
+		break;
+
+	case REPORT_LUNS:
+		ahciem_rbuf_fill(&args, ahciem_scsiop_report_luns);
+		break;
+
+	case REQUEST_SENSE:
+		ahciem_scsi_set_sense(cmd, 0, 0, 0);
+		break;
+
+	case TEST_UNIT_READY:
+		/* born ready */
+		break;
+
+	case SEND_DIAGNOSTIC:
+		if (cdb[1] & 0x10)	/* PF (page format) */
+			ahciem_sesop_txdx(enc, cmd);
+		else
+			ahciem_scsi_set_invalid_field(cmd, 1, 4, true);
+		break;
+
+	case RECEIVE_DIAGNOSTIC:
+		switch (cdb[2]) {
+		case 0x0:	/* Supported Diagnostic Pages */
+			ahciem_rbuf_fill(&args, ahciem_sesop_rxdx_0);
+			break;
+		case 0x1:	/* Configuration */
+			ahciem_rbuf_fill(&args, ahciem_sesop_rxdx_1);
+			break;
+		case 0x2:	/* Enclosure Status */
+			ahciem_rbuf_fill(&args, ahciem_sesop_rxdx_2);
+			break;
+		case 0x7:	/* Element Descriptor */
+			ahciem_rbuf_fill(&args, ahciem_sesop_rxdx_7);
+			break;
+		case 0xa:	/* Additional Element Status */
+			ahciem_rbuf_fill(&args, ahciem_sesop_rxdx_a);
+			break;
+		default:
+			ahciem_scsi_set_invalid_field(cmd, 2, 0, true);
+			break;
+		}
+		break;
+
+	default:
+		ahciem_scsi_set_sense(cmd, ILLEGAL_REQUEST, 0x20, 0x0);
+		break;
+	}
+
+	scsi_done(cmd);
```

Comment:
This line was modified. Previously, `scsi_cmnd` struct contained a function pointer for scsi_done callback. But now it is decalared as a function in scsi.h.